libs/@local/hql/README.md

```diff
@@ -0,0 +1,51 @@
+# HQL - HASH Query Language
+
+This is the compiler, planner and executor for the HASH Query Language (HQL). HQL is a functional programming language that is created using a subset of JSON.
+
+## Language
+
+Language design documentation and a specification of the language will be provided in the future.
+
+The CST of the language is very basic, only consisting of three different constructs:
+
+- Function Invocation: A function invocation has multiple possible forms:
+  - `[name, ...args]`
+  - `{"fn": name, "args": args}`
+- Constants: Constants are simple values that are not functions, they are defined using:
+  - `{"const": value}`, this will convert the JSON type to the corresponding `hql` type.
+  - `{"const": value, "type": type}`, this will convert the JSON type to the specified `hql` type (if possible), if not possible this will result in an error.
+- Variables: Variables are references to values that are defined elsewhere, they are defined using:
+  - `"symbol"` (any string is a reference to a variable)
+  - `{"var": name}` (this is a more explicit way of defining a variable)
```

Comment:
Why do we have two representation for a variable?

Reply:
same as in all other cases, mostly not needed, but I want to give the user the option to specifically specify what they want to have, in case there is an ambiguity of something e.g. being parsed as a variable, even tho it should be a signature and vice versa.

I do not forsee this happening right now, but I want to keep the option open.

In future this also will be expanded, so you're able to have:

```json
{"var": [path, to, value, to, access]}
```

(the idea is to allow the same in string representation, for easily human use, but for machines, using an array is easier)

so e.g.

```json
"foo.bar?.baz"
{"var": ["foo", "bar?", "baz"]}
```

Reply:
> for easily human use

I don't think human users will ever write CSTs. Either they use the syntax or some tools to generate the CSTs for them (e.g. parsing UI elements or in scripts). Having multiple way of doing the same things (especially from the get-go) confuses more than it helps.

Reply:
I agree to disagree here. I've had the pleasure to write quite a bit of queries using the current language AST, and I don't see that going away.

You will need _a way_ to get the data out programmatically as well, even if just exploratory, and authoring a huge tree of nodes hinders that goal.